include/onnxruntime/core/platform/threadpool.h

```diff
@@ -1,95 +1,267 @@
-#if __GNUC__ >= 6
-#pragma GCC diagnostic ignored "-Wignored-attributes"
-#endif
-#pragma GCC diagnostic ignored "-Wunused-parameter"
-#elif defined(_MSC_VER)
-#pragma warning(push)
-#pragma warning(disable : 4267)
-#pragma warning(disable : 4127)
-#endif
-#include <unsupported/Eigen/CXX11/ThreadPool>
-#if defined(__GNUC__)
-#pragma GCC diagnostic pop
-#elif defined(_MSC_VER)
-#pragma warning(pop)
-#endif
+#include <functional>
+#include <memory>
+
+// This file use PIMPL to avoid having eigen headers here
+
+namespace Eigen {
+class Allocator;
+class ThreadPoolInterface;
+struct ThreadPoolDevice;
+}  // namespace Eigen
 
 namespace onnxruntime {
 
-namespace concurrency {
+struct TensorOpCost {
+  double bytes_loaded_;
+  double bytes_stored_;
+  double compute_cycles_;
+};
 
-/**
- * Generic class for instantiating thread pools.
- * Don't put any object of this type into a global variable in a Win32 DLL.
- */
+template <typename Environment>
+class ThreadPoolTempl;
+namespace concurrency {
 class ThreadPool {
  public:
-  /*
-  Initializes a thread pool given the current environment.
-  */
-  ThreadPool(const std::string& name, int num_threads);
-
-  /*
-  Enqueue a unit of work.
-  */
+  // Scheduling strategies for ParallelFor. The strategy governs how the given
+  // units of work are distributed among the available threads in the
+  // threadpool.
+  enum class SchedulingStrategy {
+    // The Adaptive scheduling strategy adaptively chooses the shard sizes based
+    // on the cost of each unit of work, and the cost model of the underlying
+    // threadpool device.
+    //
+    // The 'cost_per_unit' is an estimate of the number of CPU cycles (or
+    // nanoseconds if not CPU-bound) to complete a unit of work. Overestimating
+    // creates too many shards and CPU time will be dominated by per-shard
+    // overhead, such as Context creation. Underestimating may not fully make
+    // use of the specified parallelism, and may also cause inefficiencies due
+    // to load balancing issues and stragglers.
+    kAdaptive,
+    // The Fixed Block Size scheduling strategy shards the given units of work
+    // into shards of fixed size. In case the total number of units is not
+    // evenly divisible by 'block_size', at most one of the shards may be of
+    // smaller size. The exact number of shards may be found by a call to
+    // NumShardsUsedByFixedBlockSizeScheduling.
+    //
+    // Each shard may be executed on a different thread in parallel, depending
+    // on the number of threads available in the pool. Note that when there
+    // aren't enough threads in the pool to achieve full parallelism, function
+    // calls will be automatically queued.
+    kFixedBlockSize
+  };
+
+  // Contains additional parameters for either the Adaptive or the Fixed Block
+  // Size scheduling strategy.
+  class SchedulingParams {
+   public:
+    explicit SchedulingParams(SchedulingStrategy strategy, optional<int64_t> cost_per_unit,
```

Comment:
>optional [](start = 59, length = 8)

Can this class with std::optional when we start using? Should we namespace qualify it?

Reply:
std::optional requires C++17 so we copied one implementation to our onnxruntime namespace. Similar story to make_unique. That's why there it doesn't have std prefix.